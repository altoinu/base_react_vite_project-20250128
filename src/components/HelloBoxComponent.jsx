import styles from "./HelloBoxComponent.module.css";
import PropTypes from "prop-types";
import { useState } from "react";

function HelloBoxComponent({ personName, onButtonClick, ...props }) {
  const [clickedButtonResponse, setClickedButtonResponse] = useState();

  const handleButtonClick = (buttonNum) => {
    const response =
      buttonNum === 1
        ? "First button clicked"
        : buttonNum === 2
          ? "Second button clicked"
          : "something else clicked";

    setClickedButtonResponse(response);

    onButtonClick?.(buttonNum);
  };

  return (
    <>
      <h1>
        Hello {personName}
        {props.secondPersonName ? ` and ${props.secondPersonName}` : ""}!
      </h1>
      {onButtonClick && (
        <div className="flex flex-row">
          <button
            className={styles.button}
            onClick={() => handleButtonClick(1)}
          >
            Click Me!
          </button>
          <button
            className={styles.button}
            onClick={() => handleButtonClick(2)}
          >
            Click Me Too!
          </button>
        </div>
      )}
      {clickedButtonResponse && (
        <span>clicked button response: {clickedButtonResponse}</span>
      )}
    </>
  );
}

// needed for ESLint rule
HelloBoxComponent.propTypes = {
  personName: PropTypes.string.isRequired,
  secondPersonName: PropTypes.string,
  onButtonClick: PropTypes.func,
};

export default HelloBoxComponent;
