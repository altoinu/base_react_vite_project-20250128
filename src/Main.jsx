import FetchDataComponent from "./components/FetchDataComponent";
import HelloBoxComponent from "./components/HelloBoxComponent";

export default function Main() {
  function handleButtonClick(buttonNum) {
    alert("button number clicked:" + buttonNum);
  }

  return (
    <>
      <h1>Hello from React!!!</h1>
      <HelloBoxComponent personName="Kaoru" />
      <br />
      <HelloBoxComponent
        personName="John"
        secondPersonName="Doe"
        onButtonClick={handleButtonClick}
      />
      <FetchDataComponent />
    </>
  );
}
