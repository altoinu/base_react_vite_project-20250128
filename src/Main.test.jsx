import Main from "./Main";
import HelloBoxComponent from "./components/HelloBoxComponent";
import { describe, expect, it } from "@jest/globals";
import { render, screen } from "@testing-library/react";

// Several different ways to mock child component
/*
jest.mock("./components/HelloBoxComponent");
*/
jest.mock("./components/HelloBoxComponent", () =>
  jest.fn(() => <div>Mock HelloBoxComponent</div>),
);
/*
jest.mock("./components/HelloBoxComponent", () =>
  jest.fn(({ personName }) => <div>Mock HelloBoxComponent {personName}</div>),
);
*/
/*
jest.mock("./components/HelloBoxComponent", () => ({
  __esModule: true,
  default: ({ personName, secondPersonName }) => (
    <div role="mock-hellobox">
      <span>Mock HelloBoxComponent</span>
      <span data-testid="personName">{personName}</span>
      {secondPersonName && (
        <span data-testid="secondPersonName">{secondPersonName}</span>
      )}
    </div>
  ),
}));
*/

jest.mock("./components/FetchDataComponent", () =>
  jest.fn(() => <div>Mock FetchDataComponent</div>),
);

describe("<Main>", () => {
  it("should render", () => {
    const { container: pageContainer } = render(<Main />);

    expect(pageContainer).toMatchSnapshot();
  });

  it("should render a heading", () => {
    render(<Main />);

    /*
    const heading = screen.getByRole("heading");
    expect(heading.textContent).toBe("Hello from React!!!");
    */

    const heading2 = screen.getByText("Hello from React!!!");
    expect(heading2).toBeInTheDocument();
    expect(heading2).toHaveRole("heading");
  });

  it("should render HelloBoxComponents", () => {
    render(<Main />);

    // Tests to see if passed props were received correctly
    expect(HelloBoxComponent).toHaveBeenCalledTimes(2);
    // undefined in second argument
    // https://dev.to/peterlidee/mocking-react-components-jest-mocking-react-part-2-2l8j
    expect(HelloBoxComponent).toHaveBeenNthCalledWith(
      1,
      expect.objectContaining({ personName: "Kaoru" }),
      undefined,
    );
    expect(HelloBoxComponent).toHaveBeenNthCalledWith(
      2,
      expect.objectContaining({ personName: "John", secondPersonName: "Doe" }),
      undefined,
    );

    /*
    // Other ways to test props, by checking actual elements in mocked component
    const helloBox = screen.getAllByRole("mock-hellobox");
    expect(helloBox).toHaveLength(2);

    expect(helloBox[0]).toBeInTheDocument();
    expect(within(helloBox[0]).queryByTestId("personName")).toHaveTextContent(
      "Kaoru",
    );
    expect(
      within(helloBox[0]).queryByTestId("secondPersonName"),
    ).not.toBeInTheDocument();

    expect(helloBox[1]).toBeInTheDocument();
    expect(within(helloBox[1]).queryByTestId("personName")).toHaveTextContent(
      "John",
    );
    expect(
      within(helloBox[1]).queryByTestId("secondPersonName"),
    ).toHaveTextContent("Doe");
    */
  });
});
