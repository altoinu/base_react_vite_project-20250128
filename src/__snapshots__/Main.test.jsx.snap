// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main> should render 1`] = `
<div>
  <h1>
    Hello from React!!!
  </h1>
  <div>
    Mock HelloBoxComponent
  </div>
  <br />
  <div>
    Mock HelloBoxComponent
  </div>
</div>
`;
